### PR TITLE
Test against external cassandra using docker

### PR DESCRIPTION
**Pull request by chbatey**
I am sure some tests will fail, just trying this out to see how fast it would be

Fixes https://github.com/akka/akka-persistence-cassandra/issues/370

### CODE REVIEW

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournalUnexpectedError.scala [hunk 1]
```diff
@@ -1,3 +1,7 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <https://www.lightbend.com>
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournalUnexpectedError.scala]: Not up to 2019?

Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournalUnexpectedError.scala]: I won't change that in this PR but needs updating

Comment (johanandren) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournalUnexpectedError.scala]: Ah, it's the plugin that did it?

File: core/src/main/scala/akka/persistence/cassandra/EventsByTagMigration.scala [hunk 2]
```diff
@@ -105,23 +108,25 @@ class EventsByTagMigration(
     with CassandraTagRecovery {
 
   private[akka] val log = Logging.getLogger(system, getClass)
-  private val queries = PersistenceQuery(system).readJournalFor[CassandraReadJournal](readJournalNamespace)
+  private lazy val queries = PersistenceQuery(system).readJournalFor[CassandraReadJournal](readJournalNamespace)
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/EventsByTagMigration.scala]: Don't want the journal to connect eagerly and do its initialisation and get the metadata as this tests changes the tables

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 3]
```diff
@@ -286,15 +286,18 @@ class CassandraJournal(cfg: Config)
...
-      // notify TagWriters when write was successful
-      result.foreach(bulkTagWrite => tagWrites.foreach(_ ! bulkTagWrite))
+    bulkTagWrite.onComplete { _ =>
+      self ! WriteFinished(pid, writeInProgressForPersistentId.future)
+      writeInProgressForPersistentId.success(Done)
     }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
 
     //Nil == all good
-    tws.map(_ => Nil)(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+    bulkTagWrite.map(bulkTagWrite => {
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: this is https://github.com/akka/akka-persistence-cassandra/pull/468 which I did as a separate PR as not to block the bug fix on getting this build stable

File: core/src/multi-jvm/scala/akka/cluster/persistence/cassandra/EventsByTagMultiJvmSpec.scala [hunk 4]
```diff
@@ -101,23 +102,26 @@ abstract class EventsByTagMultiJvmSpec extends MultiNodeSpec(EventsByTagMultiJvm
       var readers: Seq[(Int, TestSubscriber.Probe[(String, Int)])] = Nil
 
       runOn(node2, node3) {
-        readers = (0 until readersPerNode) map { i =>
-          (i, queryJournal.eventsByTag("all", NoOffset)
-            .map(e => (e.persistenceId, e.event.asInstanceOf[Int]))
-            .runWith(TestSink.probe))
+        readers = (0 until readersPerNode).map { i =>
+          (
```
Comment (chbatey) [inline core/src/multi-jvm/scala/akka/cluster/persistence/cassandra/EventsByTagMultiJvmSpec.scala]: :man_shrugging: scalafmt :-/

File: core/src/test/scala/akka/persistence/cassandra/CassandraEventsByTagLoadSpec.scala [hunk 5]
```diff
@@ -27,7 +27,7 @@ object CassandraEventsByTagLoadSpec {
        cassandra-journal {
          log-queries = off
          events-by-tag {
-            max-message-batch-size = 200
+            max-message-batch-size = 25
```
Comment (chbatey) [inline core/src/test/scala/akka/persistence/cassandra/CassandraEventsByTagLoadSpec.scala]: this produced warnings about batch size

File: core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala [hunk 6]
```diff
@@ -97,12 +98,37 @@ abstract class CassandraSpec(
...
+
+    out
+  }
   override protected def externalCassandraCleanup(): Unit =
     Try {
-      system.log.info(s"Dropping keysapces: $journalName $snapshotName")
-      cluster.connect().execute(s"drop keyspace if exists $journalName")
-      cluster.connect().execute(s"drop keyspace if exists $snapshotName")
-      cluster.close()
+      val c = cluster.connect(journalName)
```
Comment (chbatey) [inline core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala]: this dumps the rows if the tests fails, has been very useful for debugging build failures.

File: core/src/test/scala/akka/persistence/cassandra/journal/CassandraJournalDeletionSpec.scala [hunk 7]
```diff
@@ -46,10 +46,10 @@ object CassandraJournalDeletionSpec {
         }
       case GetRecoveredEvents =>
         sender() ! RecoveredEvents(recoveredEvents.reverse)
-      case DeleteTo(to) => deleteMessages(to)
+      case DeleteTo(to) =>
+        deleteMessages(to)
       case DeleteMessagesSuccess(to) =>
         context.system.log.debug("Deleted to: {}", to)
```
Comment (chbatey) [inline core/src/test/scala/akka/persistence/cassandra/journal/CassandraJournalDeletionSpec.scala]: this assert here has been removed as inside akka persistence delete confirmations to the PA are not ordered, which could be a very unimportant bug? I've confirmed that the journal does delete them in the correct order but hard to tests that explicitly

File: core/src/test/scala/akka/persistence/cassandra/journal/CassandraJournalDeletionSpec.scala [hunk 8]
```diff
@@ -61,7 +61,7 @@ object CassandraJournalDeletionSpec {
 }
 
 class CassandraJournalDeletionSpec extends CassandraSpec(s"""
-    akka.loglevel = INFO
+    akka.loglevel = DEBUG
```
Comment (chbatey) [inline core/src/test/scala/akka/persistence/cassandra/journal/CassandraJournalDeletionSpec.scala]: FIXME, switch back to DEBUG before comitting

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala [hunk 9]
```diff
@@ -492,10 +492,10 @@ class EventsByTagSpec extends AbstractEventsByTagSpec(EventsByTagSpec.config) {
 }
 
 class EventsByTagZeroEventualConsistencyDelaySpec
-    extends AbstractEventsByTagSpec(
-      ConfigFactory
-        .parseString("cassandra-query-journal.eventual-consistency-delay = 0s")
-        .withFallback(EventsByTagSpec.strictConfig)) {
+    extends AbstractEventsByTagSpec(ConfigFactory.parseString("""
+            cassandra-query-journal.eventual-consistency-delay = 0s
+            akka.loglevel = INFO
```
Comment (chbatey) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: FIXME, remove

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala [hunk 10]
```diff
@@ -534,7 +534,9 @@ class EventsByTagZeroEventualConsistencyDelaySpec
 // Manually writing events for same persistence id is no longer valid
 // tho new persistence ids from a node with a slow clocks is still applicable
 // and we pick them up by noticing that there is a tagPidSequenceNr gap
-class EventsByTagFindDelayedEventsSpec extends AbstractEventsByTagSpec(EventsByTagSpec.strictConfig) {
+class EventsByTagFindDelayedEventsSpec extends AbstractEventsByTagSpec(ConfigFactory.parseString("""
+akka.loglevel = DEBUG
```
Comment (chbatey) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: FIXME, remove

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournalUnexpectedError.scala (hunk 1 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournalUnexpectedError.scala]: Yup, we're do out of date: https://github.com/akka/akka-persistence-cassandra/blob/master/build.sbt#L53

Will raise a separate PR to fix that

**Comment by chbatey** [comment]
Omg it worked!

**Comment by chbatey** [comment]
I have a few small changes to make that i'll push as different commits so this runs many times

**Comment by chbatey** [comment]
The test problems are due to the slower response time from the docker cassandra. The speed up from this change is from not starting the C* for each tests but the C* in docker seems to be slightly slower to respond.

We had many tests that have settings that rely on quick response times e.g. setting the refresh interval for a query at 100ms and a timeout of 250ms assuming it would get two attempts but when the query takes longer it doesn't get a second attempt. These tests would also be susceptible to a GC pause.

Most of the tests I've fixed have failed very intermittently in the past.

File: core/src/main/scala/akka/persistence/cassandra/query/CassandraReadStatements.scala [hunk 11]
```diff
@@ -21,22 +21,14 @@ import akka.annotation.InternalApi
       SELECT DISTINCT persistence_id, partition_nr FROM $tableName
      """
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/query/CassandraReadStatements.scala]: Variation that doesn't include a end timestamp not used

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala [hunk 12]
```diff
@@ -534,7 +542,20 @@ class EventsByTagZeroEventualConsistencyDelaySpec
...
+class EventsByTagFindDelayedEventsSpec
+    extends AbstractEventsByTagSpec(
+      ConfigFactory
+        .parseString(
+          """
+akka.loglevel = DEBUG
+# find delayed events from offset relies on this as it puts an event before the offset that will not
+# be found and one after that will be found for a new persistence id
+# have it at least 2x the interval so searching for missing tries trice
+cassandra-query-journal.events-by-tag.new-persistence-id-scan-timeout = 600ms
```
Comment (chbatey) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: this is an important change

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala [hunk 13]
```diff
@@ -17,7 +17,9 @@ import akka.persistence.query.scaladsl.{ CurrentEventsByTagQuery, EventsByTagQue
 import akka.persistence.query.{ EventEnvelope, NoOffset, TimeBasedUUID }
 import akka.persistence.{ PersistentActor, PersistentRepr }
 import akka.serialization.SerializationExtension
+import akka.stream.testkit.TestSubscriber
```
Comment (chbatey) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: Most of this file's changes are shutting down the streams if there is a failure so the logs are spammed for the rest of the run, i'll mark out important changes

**Comment by chbatey** [comment]
This is ready for review now. I have kept one DEBUG logging on as I want to i see it pass a few more times before merging

File: core/src/multi-jvm/scala/akka/cluster/persistence/cassandra/EventsByTagMultiJvmSpec.scala (hunk 4 above)
Comment (patriknw) [reply core/src/multi-jvm/scala/akka/cluster/persistence/cassandra/EventsByTagMultiJvmSpec.scala]: maybe `i -> ...` instead of `(i,...)`

**Comment by chbatey** [comment]
Yep i did leave a comment on each important change but some of them are out dated now

The actual bug fix i pulled out as a separate PR to highlight it

File: core/src/test/scala/akka/persistence/cassandra/journal/CassandraJournalDeletionSpec.scala [hunk 14]
```diff
@@ -114,10 +116,11 @@ class CassandraJournalDeletionSpec extends CassandraSpec(s"""
...
 
-      // Should be in order, previously they could over take each other.
-      (1L to 99L).foreach { i =>
-        deleteSuccess.expectMsg(Deleted(i))
-      }
+      // The AsyncWriteJournal does not guarantee that DeleteSuccess are delivered in the order
+      // that they are completed by the journal implementation so can't assert this reliably
+      (1L to 99L).map { _ =>
+        deleteSuccess.expectMsgType[Deleted].sequenceNr
+      }.toSet shouldEqual (1L to 99L).toSet
```
[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/test/scala/akka/persistence/cassandra/journal/CassandraJournalDeletionSpec.scala]: Yep

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagStageSpec.scala [hunk 15]
```diff
@@ -30,7 +30,7 @@ object EventsByTagStageSpec {
   val today = LocalDateTime.now(ZoneOffset.UTC)
   val fetchSize = 3L
   val config = ConfigFactory.parseString(s"""
-        akka.loglevel = INFO
+        akka.loglevel = DEBUG
```
[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/test/scala/akka/persistence/cassandra/query/EventsByTagStageSpec.scala]: This is the test i am least confident about so thought i would leave it in for any review feedback travis runs

**Comment by chbatey** [comment]
I am going to merge this, whereas I am not fully convinced some of the specs are stable running in travis against docker I think they are more stable than the embedded c* per test.

**Pull request merged**